lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -63,6 +59,64 @@ struct ConnectInfo {
 using InnerRefMap =
     DenseMap<hw::InnerRefAttr, std::pair<hw::InnerSymAttr, StringAttr>>;
 
+//===----------------------------------------------------------------------===//
+// Naming Helpers
+//===----------------------------------------------------------------------===//
+
+static void appendName(StringRef name, SmallString<32> &output,
+                       bool toLower = false) {
+  if (name.empty())
+    return;
+  if (!output.empty())
+    output.append("_");
+  output.append(name);
+  if (!toLower)
+    return;
+  auto i = output.size() - name.size();
+  output[i] = llvm::toLower(output[i]);
+}
+
+static void appendName(SymbolRefAttr name, SmallString<32> &output,
+                       bool toLower = false) {
+  appendName(name.getRootReference(), output, toLower);
+  for (auto nested : name.getNestedReferences())
+    appendName(nested.getValue(), output, toLower);
+}
+
+/// For a layer `@A::@B::@C` in module Module,
+/// the generated module is called `Module_A_B_C`.
+static SmallString<32> moduleNameForLayer(StringRef moduleName,
+                                          SymbolRefAttr layerName) {
+  SmallString<32> result;
+  appendName(moduleName, result);
+  appendName(layerName, result);
+  return result;
+}
+
+/// For a layerblock `@A::@B::@C`,
+/// the generated instance is called `a_b_c`.
+static SmallString<32> instanceNameForLayer(SymbolRefAttr layerName) {
+  SmallString<32> result;
+  appendName(layerName, result, /*toLower=*/true);
+  return result;
+}
+
+/// For all layerblocks `@A::@B::@C` in a circuit called Circuit,
+/// the output filename is `layers_Circuit_A_B_C.sv`.
+static SmallString<32> fileNameForLayer(StringRef circuitName,
```

Comment:
Does this change for FIRRTL 4.0?

abi.md presently reads:

```
filename = "layers_" , module , "_", root , { "_" , nested } , ".sv" ;
```

(module not circuit)

Is this going to work for public modules that aren't the main module as-is?

Reply:
I think this pass does not do the right thing for public modules. This happens to work now for the main module. The pass should be updated to emit the collateral for all public modules in a follow-on. Good catch, Will.

Reply:
Opened an issue: https://github.com/llvm/circt/issues/6748